### PR TITLE
Refactor nox to separate integration tests

### PR DESCRIPTION
**Pull request by patrick91**
This PR removes all integrations deps from dev dependencies
and instead adds them to a new group in poetry, you can install
all like this:

```
poetry install --with integrations
```

Meanwhile on CI will be able to run tests based on the individual 
dependency. This should make it easier to tests with newer python versions

### CODE REVIEW

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to @patrick91 for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

**Comment by botberry** [comment]
## Apollo Federation Subgraph Compatibility Results

<table>
<thead>
<tr><th>Federation 1 Support</th><th>Federation 2 Support</th></tr>
</thead>
<tbody>
<tr><td><table><tr><th><code>_service</code></th><td>🟢</td></tr><tr><th><code>@key (single)</code></th><td>🟢</td></tr><tr><th><code>@key (multi)</code></th><td>🟢</td></tr><tr><th><code>@key (composite)</code></th><td>🟢</td></tr><tr><th><code>repeatable @key</code></th><td>🟢</td></tr><tr><th><code>@requires</code></th><td>🟢</td></tr><tr><th><code>@provides</code></th><td>🟢</td></tr><tr><th><code>federated tracing</code></th><td>🔲</td></tr></table></td><td><table><tr><th><code>@link</code></th><td>🟢</td></tr><tr><th><code>@shareable</code></th><td>🟢</td></tr><tr><th><code>@tag</code></th><td>🟢</td></tr><tr><th><code>@override</code></th><td>🟢</td></tr><tr><th><code>@inaccessible</code></th><td>🟢</td></tr><tr><th><code>@composeDirective</code></th><td>🟢</td></tr><tr><th><code>@interfaceObject</code></th><td>🟢</td></tr></table></td></tr>
</tbody>
</table>

Learn more:
* [Apollo Federation Subgraph Specification](https://www.apollographql.com/docs/federation/subgraph-spec/)
* [Compatibility Tests](https://github.com/apollographql/apollo-federation-subgraph-compatibility/blob/main/COMPATIBILITY.md)

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -55,20 +55,20 @@ jobs:
...
+      #     path: |
+      #       ~/.cache/pip
+      #       ~/.nox
+      #       .nox
+      #     key:
+      #       ${{ runner.os }}-nox-${{ matrix.session.session }}-${{
+      #       hashFiles('**/poetry.lock') }}-${{ hashFiles('**/noxfile.py') }}
+      #     restore-keys: |
+      #       ${{ runner.os }}-nox-${{ matrix.session.session }}-
+      #       ${{ runner.os }}-nox-
```
Comment (bellini666) [inline .github/workflows/test.yml:71]: I assume this is commented for testing the CI =P

Just a reminder to uncomment this before merging

**Review by bellini666 — APPROVED** [review]
The PR is massive but the changes are really obvious :)

LGTM 👍🏼

**Comment by bellini666** [comment]
Btw, don't know if it is codecov that is going crazy or something wrong in this change, but codecov is reporting a `-5.23%` project coverage

**Comment by patrick91** [comment]
> Btw, don't know if it is codecov that is going crazy or something wrong in this change, but codecov is reporting a `-5.23%` project coverage

this is because of the changes, I'm still not done (I thought I was) :D

File: .github/workflows/test.yml (hunk 1 above)
Comment (patrick91) [reply .github/workflows/test.yml:71]: yeah, there's something wrong with the cache, I might keep it disabled tbh, and find a proper fix later on

File: tests/asgi/test_async.py [hunk 2]
```diff
@@ -1,26 +1,33 @@
...
     @strawberry.type
     class Query:
         @strawberry.field
-        async def hello(self, name: typing.Optional[str] = None) -> str:
+        async def hello(self, name: Optional[str] = None) -> str:
             return f"Hello {name or 'world'}"
 
     async_schema = strawberry.Schema(Query)
-    app = GraphQL(async_schema)
+    app = GraphQL[None, None](async_schema)
```
Comment (skilkis) [inline tests/asgi/test_async.py:26]: Interesting is this for proper typing?

File: noxfile.py [hunk 3]
```diff
@@ -1,32 +1,57 @@
 import nox
 from nox_poetry import Session, session
 
+nox.options.reuse_existing_virtualenvs = True
+nox.options.error_on_external_run = True
+
 PYTHON_VERSIONS = ["3.11", "3.10", "3.9", "3.8", "3.7"]
 
 
+COMMON_PYTEST_OPTIONS = [
```
Comment (skilkis) [inline noxfile.py:10]: 🎉

File: tests/http/conftest.py [hunk 4]
```diff
@@ -1,58 +1,54 @@
...
-        pytest.param(
-            # SyncChannelsHttpClient uses @database_sync_to_async and therefore
-            # needs pytest.mark.django_db
-            SyncChannelsHttpClient,
-            marks=[pytest.mark.channels, pytest.mark.django_db],
+from .clients.base import HttpClient
+
+
+def _get_http_client_classes() -> Generator[Any, None, None]:
+    for client, module, marks in [
```
Comment (skilkis) [inline tests/http/conftest.py:10]: As a nitpick (Open Closed Principle) I'd move the loop outside of this function and iteratively call the body of the loop to within a generator comprehension in the params argument to the fixture

File: tests/websockets/conftest.py [hunk 5]
```diff
@@ -1,36 +1,47 @@
...
-    params=clients,
-    ids=ids,
-)
-def http_client_class(request) -> Type[HttpClient]:
+from ..http.clients.base import HttpClient
+
+
+def _get_http_client_classes() -> Generator[Any, None, None]:
+    for client, module, marks in [
+        ("AioHttpClient", "aiohttp", [pytest.mark.aiohttp]),
```
Comment (skilkis) [inline tests/websockets/conftest.py:11]: Same nitpick here if you choose to accept the other one ☺️

**Review by skilkis — APPROVED** [review]
LGTM 🎉

File: tests/asgi/test_async.py (hunk 2 above)
Comment (patrick91) [reply tests/asgi/test_async.py:26]: yup! it's unfortunate we can't have default typevars :(

Comment (patrick91) [reply tests/asgi/test_async.py:26]: yup! it's unfortunate we can't have default typevars :(

File: tests/http/conftest.py (hunk 4 above)
Comment (patrick91) [reply tests/http/conftest.py:10]: can you show me how you'd do it? 😊

Comment (patrick91) [reply tests/http/conftest.py:10]: (I'm begin lazy :P)

**Comment by patrick91** [comment]
yay!

> Merging [#2950](https://app.codecov.io/gh/strawberry-graphql/strawberry/pull/2950?src=pr&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=strawberry-graphql) (https://github.com/strawberry-graphql/strawberry/commit/cb7efcfed832222e65f23b9bd172c0c5372c3595) into [main](https://app.codecov.io/gh/strawberry-graphql/strawberry/commit/60bfd7639c612f0dbd34a14f0a75823c34802246?el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=strawberry-graphql) (https://github.com/strawberry-graphql/strawberry/commit/60bfd7639c612f0dbd34a14f0a75823c34802246) will increase coverage by 0.24%.
The diff coverage is 94.08%.

**Comment by patrick91** [comment]
yay!

> Merging [#2950](https://app.codecov.io/gh/strawberry-graphql/strawberry/pull/2950?src=pr&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=strawberry-graphql) (https://github.com/strawberry-graphql/strawberry/commit/cb7efcfed832222e65f23b9bd172c0c5372c3595) into [main](https://app.codecov.io/gh/strawberry-graphql/strawberry/commit/60bfd7639c612f0dbd34a14f0a75823c34802246?el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=strawberry-graphql) (https://github.com/strawberry-graphql/strawberry/commit/60bfd7639c612f0dbd34a14f0a75823c34802246) will increase coverage by 0.24%.
The diff coverage is 94.08%.

File: tests/http/conftest.py (hunk 4 above)
Comment (skilkis) [reply tests/http/conftest.py:10]: Hey sorry Patrick I didn't have access to my PC when writing the comment, I just saw your message now. I'll post how to do it here when I get home 😊

Comment (patrick91) [reply tests/http/conftest.py:10]: thanks! I forgot to ping you!

Comment (skilkis) [reply tests/http/conftest.py:10]: So something like this:

```python
import importlib
from typing import Any, Sequence, Type

import pytest

from .clients.base import HttpClient


def _get_http_client_class(
    client: str,
    module: str,
    marks: Sequence[pytest.MarkDecorator],
) -> Any:
    try:
        client_class = getattr(
            importlib.import_module(f".{module}", package="tests.http.clients"),
            client,
        )
    except ImportError:
        client_class = None

    return pytest.param(
        client_class,
        marks=[
            *marks,
            pytest.mark.skipif(
                client_class is None, reason=f"Client {client} not found"
            ),
        ],
    )


@pytest.fixture(
    params=(
        _get_http_client_class(client, module, marks)
        for client, module, marks in [
            ("AioHttpClient", "aiohttp", [pytest.mark.aiohttp]),
            ("AsgiHttpClient", "asgi", [pytest.mark.asgi]),
            ("AsyncDjangoHttpClient", "async_django", [pytest.mark.django]),
            ("AsyncFlaskHttpClient", "async_flask", [pytest.mark.flask]),
            ("ChannelsHttpClient", "channels", [pytest.mark.channels]),
            ("ChaliceHttpClient", "chalice", [pytest.mark.chalice]),
            ("DjangoHttpClient", "django", [pytest.mark.django]),
            ("FastAPIHttpClient", "fastapi", [pytest.mark.fastapi]),
            ("FlaskHttpClient", "flask", [pytest.mark.flask]),
            ("SanicHttpClient", "sanic", [pytest.mark.sanic]),
            ("StarliteHttpClient", "starlite", [pytest.mark.starlite]),
            (
                "SyncChannelsHttpClient",
                "channels",
                [pytest.mark.channels, pytest.mark.django_db],
            ),
        ]
    )
)
def http_client_class(request: Any) -> Type[HttpClient]:
    return request.param


@pytest.fixture()
def http_client(http_client_class: Type[HttpClient]) -> HttpClient:
    return http_client_class()
```

Or alternatively:

```python
import importlib
from typing import Any, Generator, Sequence, Tuple, Type

import pytest

from .clients.base import HttpClient


def parametrize_clients(
    specifications: Sequence[Tuple[str, str, Sequence[pytest.MarkDecorator]]],
) -> Generator[Any, None, None]:
    for client, module, marks in specifications:
        try:
            client_class = getattr(
                importlib.import_module(f".{module}", package="tests.http.clients"),
                client,
            )
        except ImportError:
            client_class = None

        yield pytest.param(
            client_class,
            marks=[
                *marks,
                pytest.mark.skipif(
                    client_class is None, reason=f"Client {client} not found"
                ),
            ],
        )


@pytest.fixture(
    params=parametrize_clients(
        [
            ("AioHttpClient", "aiohttp", [pytest.mark.aiohttp]),
            ("AsgiHttpClient", "asgi", [pytest.mark.asgi]),
            ("AsyncDjangoHttpClient", "async_django", [pytest.mark.django]),
            ("AsyncFlaskHttpClient", "async_flask", [pytest.mark.flask]),
            ("ChannelsHttpClient", "channels", [pytest.mark.channels]),
            ("ChaliceHttpClient", "chalice", [pytest.mark.chalice]),
            ("DjangoHttpClient", "django", [pytest.mark.django]),
            ("FastAPIHttpClient", "fastapi", [pytest.mark.fastapi]),
            ("FlaskHttpClient", "flask", [pytest.mark.flask]),
            ("SanicHttpClient", "sanic", [pytest.mark.sanic]),
            ("StarliteHttpClient", "starlite", [pytest.mark.starlite]),
            (
                "SyncChannelsHttpClient",
                "channels",
                [pytest.mark.channels, pytest.mark.django_db],
            ),
        ]
    )
)
def http_client_class(request: Any) -> Type[HttpClient]:
    return request.param


@pytest.fixture()
def http_client(http_client_class: Type[HttpClient]) -> HttpClient:
    return http_client_class()

```

Could then re-use the same `parametrize_clients` in the websockets conftest

**Pull request merged**